### ISSUE TITLE
Creating value types is somewhat gnarly in API

### ISSUE THREAD

**Issue by pontaoski** [jwharm/java-gi#86]
Calling `Rect.allocate(Arena.global(),Point.allocate(Arena.global(), 0, 0),Size.allocate(Arena.global(), w, h))` is a bit wordy compared to what I would want to write: `Rect(Point(0, 0), Size(w, h))` (`new` omitted because Scala)

**Comment by jwharm** [comment]
Thanks for your feedback, it's much appreciated.

Good suggestion to replace the `allocate` method with a constructor. I don't remember exactly why I opted for the factory method, but I agree that a constructor is nicer.

Don't use `Arena.global()`, it will never free the allocated memory. `Arena.ofAuto()` hooks into the garbage collector. But it's best to be a bit more explicit and create a local arena with a `Using` block:

```
Using.resource(Arena.ofConfined()) { a =>
    Rect.allocate(a, Point.allocate(a, 0, 0), Size.allocate(a, w, h))
}
// arena is now closed, no memory allocations left
```

The reason I didn't automatically default to `Arena.ofAuto()` is that the GC is too unpredictable. Nobody knows when it will run (if ever) and what it will clean. I wanted to encourage the `Using { ... }` pattern, which requires the explicit Arena parameter.

**Comment by jwharm** [comment]
I've added the value-type constructors in a new branch, and pushed a `0.9.1-SNAPSHOT` release if you want to try it out. It contains a fix for overriding protected methods as well.

Because snapshots releases are not on Maven Central, you'll need to add `https://oss.sonatype.org/content/repositories/snapshots` to the repositories in your build script.

**Comment by pontaoski** [comment]
Looks like this fixes calling `snapshot`, and trimming the `.allocate` makes it nicer.
Though, it's still not particularly useful because of #87.

**Comment by jwharm** [comment]
Yes I know,
Yesterday I created a fix for #87 as well, but didn't have time yet to test it properly.

**Comment by jwharm** [comment]
I published a new snapshot release with the fix for #87.

Regarding the arena parameter, I'm still in doubt. It would be easy to overload the constructors with one that defaults to `Arena.ofAuto()` and keep the current one for those who want to explicitly manage the deallocation. But on the other hand, I've found it very difficult to test proper deallocation of GObject instances with Cleaners (that also run during GC). Sometimes resources aren't deallocated at all. I fear that defaulting to `Arena.ofAuto()` will lead people to expect it to always work, while it will actually risk unpredictable memory leaks.
However, the Javadoc for `Arena.ofAuto()` literally says: "The regions of memory backing memory segments allocated with the automatic arena are deallocated at some unspecified time after the automatic arena (and all the segments allocated by it) becomes unreachable" so the JDK architects seem to be pretty sure that the memory will always (eventually) be deallocated. So but maybe I worry too much. Do you have an opinion about it?

**Comment by pontaoski** [comment]
> But on the other hand, I've found it very difficult to test proper deallocation of GObject instances with Cleaners (that also run during GC).
I assume you've been trying `System.gc()` and `System.runFinalization()`

Forcing the GC to reproducibly trigger in the middle of a test seems somewhat difficult; but also I don't think it's really necessary to test that a core part of Java is functioning as expected. 

I think having convenience constructors as a default is Good Enough:tm: and if there is memory that needs to be freed sooner than it is in some use case, manual arena management can work.

**Comment by jwharm** [comment]
I received a suggestion on matrix to let these value types use a confined arena internally. If we then let all the objects implement `AutoCloseable` it will work similar to the current solution, but without the `arena` parameter.

However, if we create nested values, like this:
```
Using.resource(Rect(Point(0, 0), Size(w, h))) { rect =>
   ...
}
```
`Rect` will be disposed, but `Point` and `Size` will leak.

To solve this, `Rect::close` could recursively free all nested values. But that would cause problems if those values are used afterwards, like this:

```
Using.resource(Point(0,0)) { point =>
    Using.resource(Rect(point, Size(w, h))) { rect =>
        ...
    }
    // Rect is now closed, and recursively closed the embedded Point and Size
    ...
    foo(point.x) <-- null dereference
}
```

So I'll stick with the current design, but I will add a constructor overload that defaults to `Arena.ofAuto()`.

To keep the API logical, in the current constructor, I will move the `Arena` parameter to the end of the argument list. So there will be three constructors:
* `Rect(Point, Size, Arena)` to specify the arena
* `Rect(Point, Size)` defaults to `Arena.ofAuto()`
* `Rect(Arena)` to allocate an empty instance with the specified Arena (already exists, will be kept unchanged)

**Comment by jwharm** [comment]
I'm adding a constructor with no parameters as well. This will allocate an uninitialized object with an auto arena.

The constructors will not be generated if there's already one or more constructors available in the introspection data.

**Comment by jwharm** [comment]
To prevent name clashes, the constructors will **not** be generated if there's already one or more constructors available in the introspection data.
Coincidentally, this impacts `Point` and `Size`, because they already have constructors: `graphene_point_alloc` and `graphene_size_alloc`, coupled with `..._init` methods to set their values.
This means that Point and Size should be created in Java with `Point.alloc().init(x, y)` etc.
I don't think it's a good idea to add more constructors for these types. It would be confusing, and it will cause name clashes with constructors named `..._new`.
Rect, on the other hand, doesn't offer its own constructor, so for Rect the new constructors will be generated.

**Comment by jwharm** [comment]
Disregard my last comment; I changed the check for existing constructors, to only match constructors named `new`. So `Point.alloc()` will happily coexist with the additional constructors.
